lib/Core/SQLiteBuildDB.cpp

```diff
@@ -373,32 +419,111 @@ class SQLiteBuildDB : public BuildDB {
+    result = sqlite3_bind_int64(findKeyForKeyIDStmt, /*index=*/1,
+                               keyID);
     assert(result == SQLITE_OK);
-    result = sqlite3_clear_bindings(findIDForKeyInRuleResultsStmt);
+    
+    KeyType key{};
+    
+    result = sqlite3_step(findKeyForKeyIDStmt);
+    if (result == SQLITE_DONE) {
+      abort();
+    } else if (result == SQLITE_ROW) {
+      assert(sqlite3_column_count(findKeyForKeyIDStmt) == 1);
+      return std::string(
+                (const char*)sqlite3_column_text(findKeyForKeyIDStmt, 0),
+                sqlite3_column_bytes(findKeyForKeyIDStmt, 0));
+    }
+    abort();
+  }
+  
+  //Finds keyID for a key and puts result in keyID_out
+  //returns false if key not present
+  bool findKeyIDForKey(const KeyType& key, uint64_t* keyID_out) {
+    int result;
+    result = sqlite3_reset(findKeyIDForKeyStmt);
+    assert(result == SQLITE_OK);
+    result = sqlite3_clear_bindings(findKeyIDForKeyStmt);
     assert(result == SQLITE_OK);
-    result = sqlite3_bind_text(findIDForKeyInRuleResultsStmt, /*index=*/1,
-                               rule.key.data(), rule.key.size(),
+    result = sqlite3_bind_text(findKeyIDForKeyStmt, /*index=*/1,
+                               key.data(), key.size(),
                                SQLITE_STATIC);
     assert(result == SQLITE_OK);
-
-    uint64_t ruleID = 0;
-    result = sqlite3_step(findIDForKeyInRuleResultsStmt);
+    
+    *keyID_out = 0;
+    result = sqlite3_step(findKeyIDForKeyStmt);
     if (result == SQLITE_DONE) {
-      ruleID = 0;
+      *keyID_out = 0;
+      return false;
     } else if (result == SQLITE_ROW) {
-      assert(sqlite3_column_count(findIDForKeyInRuleResultsStmt) == 1);
-      ruleID = sqlite3_column_int64(findIDForKeyInRuleResultsStmt, 0);
+      assert(sqlite3_column_count(findKeyIDForKeyStmt) == 1);
+      *keyID_out = sqlite3_column_int64(findKeyIDForKeyStmt, 0);
+      return true;
     } else {
       abort();
     }
+    
+    return false;
+  }
+  
+  //Adds a key into keys table if not present
+  //returns the keyID of the key
+  uint64_t addKey(const KeyType& key) {
+    int result;
+    result = sqlite3_reset(insertIntoKeysStmt);
+    assert(result == SQLITE_OK);
+    result = sqlite3_clear_bindings(insertIntoKeysStmt);
+    assert(result == SQLITE_OK);
+    result = sqlite3_bind_text(insertIntoKeysStmt, /*index=*/1,
+                               key.data(), key.size(),
+                               SQLITE_STATIC);
+    result = sqlite3_step(insertIntoKeysStmt);
+    if (result != SQLITE_DONE)
+      abort();
+    
+    uint64_t keyID;
+    auto findResult = findKeyIDForKey(key, &keyID);
+    assert(findResult == true);
+    return keyID;
+  }
+  
+  virtual void setRuleResult(const Rule& rule,
+                             const Result& ruleResult) override {
+    int result;
+
+    uint64_t keyID;
+    uint64_t ruleID = 0;
+    
+    //only bother to lookup ruleID if the key is present
```

Comment:
Nit: complete sentences (leading caps and trailing punctuation, a la the [LLVM style guide](http://llvm.org/docs/CodingStandards.html))